File: 0401-0500/0439-Ternary Expression Parser/0439-Ternary Expression Parser.cpp
class Solution {
public:
    string parseTernary(string expression) {
        stack<char> St;
        for (int i = expression.size() - 1; i >= 0; --i) {
            char c = expression[i];
            if (!St.empty() && St.top() == '?') {
                St.pop(); // ?
                char left = St.top();
                St.pop();
                char right = St.top();
                St.pop();
                if (c == 'T') {
                    St.push(left);
                }
                else {
                    St.push(right);
                }
            }
            else if (c != ':') {
                St.push(c);
            }
        }
        
        return string(1, St.top());
    }
};

File: 0401-0500/0439-Ternary Expression Parser/0439-Ternary Expression Parser.py
class Solution:
    def parseTernary(self, expression: str) -> str:
        St = []
        for i in range(len(expression) - 1, -1, -1):
            c = expression[i]
            if St and St[-1] == '?':
                St.pop()
                left = St.pop()
                right = St.pop()
                if c == 'T':
                    St.append(left)
                else:
                    St.append(right)
            elif c != ':':
                St.append(c)
        return St[0]
